### PR TITLE
A brand new conservative ordered lock table is coming.

### PR DESCRIPTION
**Pull request by pin-yu**
FifoOrderedLockTable is a new implementation of conservative locking protocol.
It is no longer stall threads on a shared objects and call notifyAll.

### CODE REVIEW

File: src/main/java/org/elasql/procedure/calvin/CalvinStoredProcedure.java [hunk 1]
```diff
@@ -53,6 +54,7 @@
 
 	private ExecutionPlan execPlan;
 	private Transaction tx;
+	private FifoLock myFifoLock = new FifoLock(txNum);
```
Comment (SLMT) [inline src/main/java/org/elasql/procedure/calvin/CalvinStoredProcedure.java]: It seems like this is not used in anywhere else. Maybe you should remove this code?

File: src/main/java/org/elasql/procedure/naive/NaiveStoredProcedure.java [hunk 2]
```diff
@@ -43,6 +44,7 @@
 	// Record keys
 	private List<PrimaryKey> readKeys = new ArrayList<PrimaryKey>();
 	private List<PrimaryKey> writeKeys = new ArrayList<PrimaryKey>();
+	private FifoLock myFifoLock = new FifoLock(txNum);
```
Comment (SLMT) [inline src/main/java/org/elasql/procedure/naive/NaiveStoredProcedure.java]: Like the one in `CalvinStoredProcedure`. This is not used.

File: src/main/java/org/elasql/storage/tx/concurrency/ConservativeOrderedCcMgr.java [hunk 3]
```diff
@@ -43,89 +46,101 @@ public ConservativeOrderedCcMgr(long txNumber) {
...
-		for (Object obj : readObjs)
-			if (!writeObjs.contains(obj))
-				lockTbl.sLock(obj, txNum);
+
+		HashSet<Object> hasLockedSet = new HashSet<Object>();
+
+		for (Object obj : writeObjs) {
+			if (hasLockedSet.contains(obj)) {
+				throw new RuntimeException("An key is tried to be locked twice");
+			}
```
Comment (SLMT) [inline src/main/java/org/elasql/storage/tx/concurrency/ConservativeOrderedCcMgr.java]: Could you explain when this happens? Is it actually an unreachable code? Or, maybe it is just for debugging?

File: src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLock.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
+package org.elasql.storage.tx.concurrency.fifolocker;
```
Comment (SLMT) [inline src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLock.java]: I suggest to put this class back to `org.elasql.storage.tx.concurrency`.

File: src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLockers.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
+package org.elasql.storage.tx.concurrency.fifolocker;
```
Comment (SLMT) [inline src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLockers.java]: I suggest to put this class back to `org.elasql.storage.tx.concurrency`.

File: src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLockers.java [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+	private PrimaryKey key;
+
+	public FifoLockers(PrimaryKey key) {
+		this.key = key;
+	}
+
+	private boolean sLocked() {
+		// avoid using sLockers.size() due to the cost of traversing the queue
+		return !sLockers.isEmpty();
+	}
```
Comment (SLMT) [inline src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLockers.java]: Suggest to remove this unused code.

File: src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLockers.java [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+		}
+
+		synchronized (nextFifoLock) {
+			sLockers.remove(myTxNum);
+			nextFifoLock.notifyLock();
+		}
+	}
+
+	public void releaseXLock(FifoLock myFifoLock) {
+		long myTxNum = myFifoLock.getTxNum();
```
Comment (SLMT) [inline src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLockers.java]: Unused code

File: src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLockers.java [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+ * record, and to keep those transactions which are waiting for the record.
+ * 
+ * @author Pin-Yu Wang
+ *
+ */
+public class FifoLockers {
+	private ConcurrentLinkedDeque<FifoLock> requestQueue = new ConcurrentLinkedDeque<FifoLock>();
+	private ConcurrentLinkedDeque<Long> sLockers = new ConcurrentLinkedDeque<Long>();
+	private AtomicLong xLocker = new AtomicLong(-1);
+	private PrimaryKey key;
```
Comment (SLMT) [inline src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLockers.java]: Unused code

**Review by SLMT — CHANGES_REQUESTED** [review]
Good! It looks well written and efficient. We only need to remove some unused code, and then we can proceed.

File: src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLockers.java (hunk 6 above)
Comment (pin-yu) [reply src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLockers.java]: Not this one though XD, the code has been removed.

File: src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLock.java (hunk 4 above)
Comment (pin-yu) [reply src/main/java/org/elasql/storage/tx/concurrency/fifolocker/FifoLock.java]: Roger that

File: src/main/java/org/elasql/storage/tx/concurrency/ConservativeOrderedCcMgr.java (hunk 3 above)
Comment (pin-yu) [reply src/main/java/org/elasql/storage/tx/concurrency/ConservativeOrderedCcMgr.java]: It's for debugging.

File: src/main/java/org/elasql/procedure/calvin/CalvinStoredProcedure.java (hunk 1 above)
Comment (pin-yu) [reply src/main/java/org/elasql/procedure/calvin/CalvinStoredProcedure.java]: I've removed

File: src/main/java/org/elasql/storage/tx/concurrency/ConservativeOrderedCcMgr.java (hunk 3 above)
Comment (pin-yu) [reply src/main/java/org/elasql/storage/tx/concurrency/ConservativeOrderedCcMgr.java]: I'll comment these line for performance optimization and give them a title "XXX: debugging"

**Pull request merged**